osc-server/src/main/java/org/osc/core/broker/util/crypto/X509TrustManagerFactory.java

```diff
@@ -220,31 +224,71 @@ private void reloadTrustManager() throws Exception {
             certificate = (X509Certificate) cf.generateCertificate(inputStream);
         }
+        setOrUpdateCertificate(certificate, newAlias);
+    }
+
+    @Override
+    public void addEntry(X509Certificate certificate, String newAlias) throws Exception {
+        if (fingerprintNotExist(getSha1Fingerprint(certificate))) {
+            setOrUpdateCertificate(certificate, newAlias);
+        } else {
+            throw new Exception("Given certificate fingerprint already exists in trust store");
+        }
+    }
+
+    public void setOrUpdateCertificate(X509Certificate certificate, String newAlias) throws Exception {
+
+        if (INTERNAL_ALIAS.equals(newAlias)) {
+            throw new Exception("Certificate alias " + INTERNAL_ALIAS + " is reserved! Change alias supplied or file name.");
+        }
+
         this.keyStore.setCertificateEntry(newAlias, certificate);
         try (FileOutputStream outputStream = new FileOutputStream(TRUSTSTORE_FILE)) {
             this.keyStore.store(outputStream, getTruststorePassword());
         }
         reloadTrustManager();
         notifyTruststoreChanged();
+
     }
 
     @Override
-    public void addEntry(X509Certificate certificate, String newAlias) throws Exception {
-        if (fingerprintNotExist(getSha1Fingerprint(certificate))) {
-            this.keyStore.setCertificateEntry(newAlias, certificate);
-            try (FileOutputStream outputStream = new FileOutputStream(TRUSTSTORE_FILE)) {
-                this.keyStore.store(outputStream, getTruststorePassword());
+    public void replaceInternalCertificate(File file, String alias, String certPass, String storePass) throws Exception {
+        KeyStore internalCertKeeper  = KeyStore.getInstance(KEYSTORE_TYPE);
+
+        Certificate[] internalCertificateChain;
+        Key internalKey;
+
+        try (FileInputStream inputStream = new FileInputStream(file)) {
+            internalCertKeeper.load(inputStream, storePass.toCharArray());
+            try {
+                 internalKey = internalCertKeeper.getKey(alias, certPass.toCharArray());
+                 internalCertificateChain = internalCertKeeper.getCertificateChain(alias);
+                 this.keyStore.setKeyEntry(INTERNAL_ALIAS, internalKey, getTruststorePassword(), internalCertificateChain);
+            } catch (ClassCastException e) {
+                throw new Exception("Internal certificate file does not contain expected certificate!", e);
+            } catch (KeyStoreException e) {
+                throw new Exception("Internal certificate file is corrupted!", e);
             }
-            reloadTrustManager();
-            notifyTruststoreChanged();
-        } else {
-            throw new Exception("Given certificate fingerprint already exists in trust store");
+        }
+
+        try {
+            this.keyStore.setKeyEntry("internal", internalKey, getTruststorePassword(), internalCertificateChain);
```

Comment:
INTERNAL_ALIAS

Reply:
jks is not necessarily the mechanism. A key pair entry, however, also contains the certificate path,which is required. As of this morning I thought they would already have a jks or p12. But it looks like a better way would be to supply a private key and a certificate.